File: Medium/200. Number of Islands/NumberOfIslands.java
/*
LeetCode Medium: 200. Number of Islands
https://leetcode.com/problems/number-of-islands/description/

Check NumberOfIslands.cpp for brief solution explanation.
*/
class Solution {
    public int numIslands(char[][] grid) {
        int M = grid.length, N = grid[0].length;
        int[][]groups = new int[M][N];
        int g = 0;
        for(int i = 0;i<M;i++){
            for(int j = 0;j<N;j++){
                if(groups[i][j]!=0) continue;
                if(groups[i][j]==0&&grid[i][j]=='1'){
                    //Found ungrouped land
                    g++;
                    dfs(grid, groups, i, j, g, M, N);
                }
            }
        }
        return g;
    }
    void dfs(char[][]grid, int[][]groups, int r, int c, int group, int M, int N){
        groups[r][c] = group;
        int[]x = {0, 0, -1, 1};
        int[]y = {1, -1, 0, 0};
        for(int i = 0;i<4;i++){
            int nxtr = r+x[i];
            int nxtc = c+y[i];
            boolean check = nxtr<M&&nxtc<N&&0<=nxtr&&0<=nxtc;
            if(check&&groups[nxtr][nxtc]==0&&grid[nxtr][nxtc]=='1'){
                dfs(grid, groups, nxtr, nxtc, group, M, N);
            }
        }
    }
}

File: Medium/200. Number of Islands/NumberOfIslands.py
'''
LeetCode Medium: 200. Number of Islands
https://leetcode.com/problems/number-of-islands/description/

Check NumberOfIslands.cpp for brief solution explanation.
'''
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        M = len(grid)
        N = len(grid[0])
        groups = [[0 for i in range(N)] for i in range(M)]
        g = 0
        x = [0, 0, 1, -1]
        y = [1, -1, 0, 0]
        def dfs(r, c, group):
            groups[r][c] = group
            for i in range(4):
                nxtr = r+x[i]
                nxtc = c+y[i]
                check = nxtr<M and nxtc<N and 0<=nxtr and 0<=nxtc
                if check and groups[nxtr][nxtc]==0 and grid[nxtr][nxtc]=='1':
                    dfs(nxtr, nxtc, group)

        for i in range(M):
            for j in range(N):
                if groups[i][j]!=0:
                    continue
                if groups[i][j]==0 and grid[i][j]=='1':
                    g+=1
                    dfs(i, j, g)
        return g